system/jlib/jtrace.hpp

```diff
@@ -0,0 +1,147 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2022 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+
+
+#ifndef JTRACE_HPP
+#define JTRACE_HPP
+
+/*
+  To use feature-level tracing flags, protect the tracing with a test such as:
+  
+  if (doTrace(flagName [, detailLevel])
+      CTXLOG("tracing regarding the specified feature...");
+
+  If detailLevel is omitted it defaults to TraceFlags::Standard. Higher detail levels can be used for feature-level tracing that is
+  less-often required.
+
+  For logging that is NOT controlled by a feature flag, but which does want to be suppressable (e.g in a "quiet" mode), you can use
+
+  if (doTrace(TraceFlags::All, TraceFlags::Standard))
+     ...
+
+  Such logging will always appear unless the detail level has been set to "None", to suppress all logging.
+
+  Feature-level trace flags are maintained per-thread, and inherited from the parent thread when a thread is created. Thus it should be
+  simple to allow tracing flags or levels to be modified for a single query or activity.
+
+*/
+
+enum class TraceFlags : unsigned
+{
+    // The trace levels apply within a feature trace setting (or globally if used with TraceFlags::All) to adjust verbosity
+    None = 0,
+    Standard = 1,
+    Detailed = 2,
+    Max = 3,
+    LevelMask = 0x3,
+    // Individual bit values above low two bits may be given meanings by specific engines or contexts, using constexpr to rename the following
+    All = 0,         // Use with doTrace to test JUST the level, without a feature flag
+    flag1 = 0x4,
+    flag2 = 0x8,
+    flag3 = 0x10,
+    flag4 = 0x20,
+    flag5 = 0x40,
+    flag6 = 0x80,
+    flag7 = 0x100,
+    flag8 = 0x200,
+    flag9 = 0x400,
+    flag10 = 0x800,
+    flag11 = 0x1000,
+    flag12 = 0x2000,
+    flag13 = 0x4000,
+    flag14 = 0x8000,
+};
+BITMASK_ENUM(TraceFlags);
+
+// Feature trace flags for hthor, thor and Roxie engines
+
+// Common to several engines
+constexpr TraceFlags traceHttp = TraceFlags::flag1;
+
+// Specific to Roxie
+constexpr TraceFlags traceRoxieLock = TraceFlags::flag2;
```

Comment:
I also was expecting component specific feature flag definitions to be shared with the component code not global.

Reply:
We are not persisting any of the flag values anywhere, so we can renumber at-will...

Reply:
From comment in TraceFlags definition:
>    // Individual bit values above low two bits may be given meanings by specific engines or contexts, using constexpr to rename the following

I think what Gavin is saying is that all feature bit flags should be defined here, i.e. all esp trace feature flags should be defined in this list too (but I may have misunderstood).
If so, I'm not what the point of TraceFlags::flag* is vs the constexpr list, why not just name in the enum list definition directly.

I had thought (and the above comment seemed to confirm it) that the idea was that common shared features flags would be listed globally, but that an area of the flag bit set would be reserved for non-common code, and could overlap, because they would never be used/have meaning at the same time, e.g. top N bits for esp features flags vs thor flags vs dali flags.

Given we have 10 feature flags defined for Roxie alone so far, I think we will run out of room quite quickly if all global.

Whether we do or don't have a section for common and a reserved bitmasked area for features flags that can overlap, should we make the thread local traceFlags var. a 64bit to give more room?